### PR TITLE
[None][fix] don't signal PDL launch completion before writing output in tinnyGemm

### PR DESCRIPTION
**Pull request by roborluo**
## Summary
The MoE-gate kernel `tinygemm2` calls `cudaTriggerProgrammaticLaunchCompletion()` **before** it writes its result (`output[...]`, the router logits). Because the kernel is launched with Programmatic Dependent Launch (PDL) enabled, the downstream PDL consumer — the MoE top-k routing kernel `customMoeRoutingKernel` — has its `cudaGridDependencySynchronize()` released by that early trigger and then reads the router logits. As a result the routing kernel can read router logits that `tinygemm2` has not finished writing → wrong top-k expert selection → **silently wrong (coherent but incorrect) outputs** under CUDA-graph generation decode.

The failure is severe in the common small-decode-batch regime (single activation batch-tile, `grid.y == ceil(batch / TILE_N=8) == 1`): for gpt-oss-120b on GB300/SM103, AIME-25 (8 repeats) drops from ~0.80 to ~0.40. It is not a crash, NaN, or truncation, so it silently fails accuracy.

## Root cause (code path)
- Producer: `cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_cuda.cu` enables PDL (`programmaticStreamSerializationAllowed = 1`); `cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh` (pre-fix) calls `cudaTriggerProgrammaticLaunchCompletion()` in the compute-warp epilogue **before** the `output[...]` global stores that follow.
- Consumer (gpt-oss path, verified): `RenormalizeMoeRoutingMethod.apply` (`tensorrt_llm/_torch/modules/fused_moe/routing.py:457-464`, taken because `num_experts(128) <= 128`, `top_k(4) <= 8`, `force_enable_pytorch_op` false) → `torch.ops.trtllm.renorm_moe_routing_op` → `renorm_moe_routing_op` (`cpp/tensorrt_llm/thop/customMoeRoutingOp.cpp:113`) → `invokeCustomMoeRouting` (`customMoeRoutingKernels.cu:225`) → `customMoeRoutingKernel`. That kernel is launched with PDL (`customMoeRoutingKernels.cu:260`, `programmaticStreamSerializationAllowed = getEnvEnablePDL()`), calls `cudaGridDependencySynchronize()` (`:128`), then reads `routerLogits[...]` (`:147`).
- Precise claim: the early trigger in `tinygemm2` can release `customMoeRoutingKernel`'s `cudaGridDependencySynchronize()` **before** `tinygemm2`'s post-trigger `output`/router-logit stores are visible, allowing a stale read. (The producer-before-store ordering and the consumer's PDL-gated read are code-verified; the read-before-store timing is supported by the accuracy collapse + recovery below, not by cycle-level instrumentation.)

## Why it depends on decode batch size (`grid.y`)
`grid = (ceil(out_features/TILE_M), ceil(batch/TILE_N=8))`; for the gpt-oss gate `out_features=128` so only `grid.y` varies with batch. The trigger-vs-store ordering is a race; with a single batch-tile (`grid.y=1`, tiny gate, the common high-fanout decode case of 2–4 tokens/server) the consumer's read lands before the stores nearly every iteration → catastrophic. With `grid.y>=2` the timing hides it.

## The fix
Remove the explicit `cudaTriggerProgrammaticLaunchCompletion()` from the epilogue. With no explicit trigger, programmatic launch completion is signaled **implicitly after all thread blocks exit** — i.e. after the `output` stores — so the consumer cannot be released before the logits are written. The kernel still participates in PDL as a **secondary** (it is launched with the PDL attribute and waits on its own producer via `cudaGridDependencySynchronize()`); only the unsafe **early release to its consumer** is removed.

This is the minimal, low-risk form. The data dependency (the routing kernel consumes `tinygemm2`'s output) means there is no safe point to trigger early anyway: the output is the last thing the kernel produces, so the earliest correct trigger is effectively at block exit, which the implicit completion already provides.

Diff: 1 file, `cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh` (remove the early trigger; add an explanatory comment).

# Test / validation evidence
gpt-oss-120b, GB300/SM103, non-disagg IFB, 72×TP1, AIME-25 ×8 (240 samples), CUDA-graph decode, PDL on unless noted. Configs verified via the engine `CudaGraphConfig(...)` echo in server logs.

Single-bucket sweep (only `cuda_graph_config.batch_sizes` changes), `grid.y = ceil(bucket/8)`:

| bucket | grid.y | AIME-25 | |
|---|---|---|---|
| `[8]`  | 1 | 0.404 | ❌ (reproduces the bug) |
| `[16]` | 2 | 0.825 | ✅ |
| `[32]` | 4 | 0.767 | ✅ |
| `[64]` | 8 | 0.854 | ✅ |

Controls (multi-bucket, full graphs, nw72):
- Stock (PDL on): AIME-25 0.433 ❌
- A no-PDL build of `tinygemm2` (launch attr removed + device PDL calls neutralized): 0.804 ✅ — isolates PDL as the cause.

Exact-fix validation (this PR's change, built as a drop-in gate op: PDL launch attr kept, `cudaGridDependencySynchronize` kept, only the early trigger removed), run at the failing regime `batch_sizes=[8]` (grid.y=1), PDL on, nw72:
- **AIME-25 = ‹FILL: deltrigger8 result, job 2238523 — got~0.80; stock at this config = 0.404›**

(240 samples ⇒ binomial SE ≈ 0.026, so 0.767/0.80/0.825/0.854 are mutually indistinguishable; 0.404 clearly is.)

# Scope / caveats
- This change removes only `tinygemm2`'s outgoing early-completion signal; it does not change the kernel math (pure control/ordering change, no numerical difference).
- A separate concern (not in this PR): `tinygemm2_cuda.cu:81-83` hard-enables PDL regardless of `TRTLLM_ENABLE_PDL`/`getEnvEnablePDL()`. Consider honoring the global switch in a follow-up.
- Reproduced on GB300/SM103; the kernel selection (`use_tinygemm`) is gated to SM 90/100/103 

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved result consistency for certain GPU execution paths by ensuring outputs are fully written before completion is signaled.
  * Helped prevent downstream consumers from seeing stale values during decode and similar workloads.

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh [hunk 1]
```diff
@@ -421,10 +421,14 @@ __global__ __launch_bounds__(384, 1) void tinygemm_kernel(__nv_bfloat16* output,
...
-            cudaTriggerProgrammaticLaunchCompletion();
-        }
+        // NOTE: Do NOT call cudaTriggerProgrammaticLaunchCompletion() here. This kernel writes its
+        // result (the `output[...]` global stores below) AFTER this point, so signaling programmatic
+        // launch completion now would let a PDL-dependent consumer proceed and read `output` before
+        // it is written -> stale gate logits / wrong MoE routing under CUDA-graph decode (the bug is
+        // severe when grid.y==1, i.e. small decode batch). With no explicit trigger, programmatic
+        // launch completion is signaled implicitly after all thread blocks exit (i.e. after the
+        // stores), which is correct. The kernel still benefits from PDL as a secondary (it overlaps
+        // weight loads / waits on the activation producer via cudaGridDependencySynchronize above).
```
Comment (dongfengy) [inline cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh:432]: Approving as this is a safe change. But the comment and the description need to clarify that calling cudaTriggerProgrammaticLaunchCompletion here is not wrong. The true root cause is unclear.

**Review by dongfengy — APPROVED** [review]

File: cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh (hunk 1 above)
Comment (roborluo) [reply cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh:432]: I think the root cause is clear enough me. It is exactly as what I described in the above. Please let me know anywhere is unlcear to you. I can provide more info if needed.

Comment (dongfengy) [reply cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh:432]: You can refer to Brian's comment on Slack. cudaTriggerProgrammaticLaunchCompletion is launch hint only. The next kernel's cudaGridDependencySynchronize will wait until all the writes of this kernel done. So the argument in this comment doesn't hold.

Comment (roborluo) [reply cpp/tensorrt_llm/kernels/tinygemm2/tinygemm2_kernel.cuh:432]: Got it. Thanks for the teaching!

**Comment by roborluo** [comment]
Close. see new PR https://github.com/NVIDIA/TensorRT-LLM/pull/15790
